service/history/queues/executable_dlq_observer.go

```diff
@@ -0,0 +1,110 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package queues
+
+import (
+	"errors"
+
+	"go.temporal.io/server/common/clock"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/service/history/tasks"
+)
+
+// ExecutableDLQObserver records telemetry like metrics and logs for an ExecutableDLQ.
+type ExecutableDLQObserver struct {
+	Executable
+	numHistoryShards  int
+	namespaceRegistry namespace.Registry
+	timeSource        clock.TimeSource
+	logger            log.Logger
+	metricsHandler    metrics.Handler
+
+	isDLQ bool
+}
+
+func NewExecutableDLQObserver(
+	executableDLQ Executable,
+	numHistoryShards int,
+	namespaceRegistry namespace.Registry,
+	timeSource clock.TimeSource,
+	logger log.Logger,
+	metricsHandler metrics.Handler,
+) *ExecutableDLQObserver {
+	return &ExecutableDLQObserver{
+		Executable:        executableDLQ,
+		numHistoryShards:  numHistoryShards,
+		namespaceRegistry: namespaceRegistry,
+		timeSource:        timeSource,
+		logger:            logger,
+		metricsHandler:    metricsHandler,
+		isDLQ:             false,
+	}
+}
+
+func (o *ExecutableDLQObserver) Execute() error {
```

Comment:
Not related to this PR but it just occurred to me.

If the State of the underlying task is not Pending (e.g. queue is shutting down or multi-cursor impl want to offload/delete some tasks from memory) anymore, the execution should be skipped.
https://github.com/temporalio/temporal/blob/main/service/history/queues/executable.go#L188